internal/controlplane/handlers_profile.go

```diff
@@ -575,3 +560,376 @@ func (s *Server) GetProfileStatusByProject(ctx context.Context,
+		if errors.As(err, &violation) {
+			log.Printf("error validating rule: %v", violation)
+			return nil, util.UserVisibleError(codes.InvalidArgument,
+				"profile contained invalid rule '%s': %s", violation.RuleType, violation.Err)
+		}
+
+		log.Printf("error getting rule type: %v", err)
+		return nil, status.Errorf(codes.Internal, "error creating profile")
+	}
+
+	oldProfile, err := getProfilePBFromDB(ctx, oldDBProfile.ID, entityCtx, qtx)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) || strings.Contains(err.Error(), "not found") {
+			return nil, util.UserVisibleError(codes.NotFound, "profile not found")
+		}
+
+		return nil, status.Errorf(codes.Internal, "failed to get profile: %s", err)
+	}
+
+	oldRules, err := s.getRulesFromProfile(ctx, oldProfile, entityCtx)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, util.UserVisibleError(codes.NotFound, "profile not found")
+		}
+		return nil, status.Errorf(codes.Internal, "error fetching profile to be updated: %v", err)
+	}
+
+	// Update top-level profile db object
+	profile, err := qtx.UpdateProfile(ctx, db.UpdateProfileParams{
+		ID:        oldDBProfile.ID,
+		Remediate: validateActionType(in.GetRemediate()),
+		Alert:     validateActionType(in.GetAlert()),
+	})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "error updating profile: %v", err)
+	}
+
+	// Create entity rules entries
+	for ent, entRules := range map[minderv1.Entity][]*minderv1.Profile_Rule{
+		minderv1.Entity_ENTITY_REPOSITORIES:       in.GetRepository(),
+		minderv1.Entity_ENTITY_ARTIFACTS:          in.GetArtifact(),
+		minderv1.Entity_ENTITY_BUILD_ENVIRONMENTS: in.GetBuildEnvironment(),
+		minderv1.Entity_ENTITY_PULL_REQUESTS:      in.GetPullRequest(),
+	} {
+		if err := updateProfileRulesForEntity(ctx, ent, &profile, qtx, entRules, rules, oldRules); err != nil {
+			return nil, err
+		}
+	}
+
+	for _, rule := range oldRules {
+		// get entity profile
+		log.Printf("getting profile for entity %s", rule.Entity)
+		entProf, err := qtx.GetProfileForEntity(ctx, db.GetProfileForEntityParams{
+			ProfileID: profile.ID,
+			Entity:    entities.EntityTypeToDB(rule.Entity),
+		})
+		if err != nil {
+			log.Printf("error getting profile for entity %s: %v", rule.Entity, err)
+			return nil, status.Errorf(codes.Internal, "error creating profile")
+		}
+
+		log.Printf("deleting rule instantiation for rule %s for entity profile %s", rule.RuleID, entProf.ID)
+		if err := qtx.DeleteRuleInstantiation(ctx, db.DeleteRuleInstantiationParams{
+			EntityProfileID: entProf.ID,
+			RuleTypeID:      rule.RuleID,
+		}); err != nil {
+			log.Printf("error deleting rule instantiation: %v", err)
+			return nil, status.Errorf(codes.Internal, "error creating profile")
+		}
+
+		log.Printf("deleting rule evaluations for rule %s and profile %s", rule.RuleID, profile.ID)
+		// Note this is done outside of the transaction to ensure we don't
+		// clash with a running evaluation.
```

Comment:
this comment is confusing, we're still in a transaction aren't we?

Reply:
now it's part of the transaction.